dot/core/process.go

```diff
@@ -0,0 +1,281 @@
+
+	// get sub-chain of block hashes
+	subchain, err := s.blockState.SubChain(startHash, endHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(subchain) > int(maxResponseSize) {
+		subchain = subchain[:maxResponseSize]
+	}
+
+	log.Trace("[core] subchain", "start", subchain[0], "end", subchain[len(subchain)-1])
+
+	responseData := []*types.BlockData{}
+
+	for _, hash := range subchain {
+
+		blockData := new(types.BlockData)
+		blockData.Hash = hash
+
+		// set defaults
+		blockData.Header = optional.NewHeader(false, nil)
+		blockData.Body = optional.NewBody(false, nil)
+		blockData.Receipt = optional.NewBytes(false, nil)
+		blockData.MessageQueue = optional.NewBytes(false, nil)
+		blockData.Justification = optional.NewBytes(false, nil)
+
+		// header
+		if (blockRequest.RequestedData & 1) == 1 {
+			retData, err := s.blockState.GetHeader(hash)
+			if err == nil && retData != nil {
+				blockData.Header = retData.AsOptional()
+			}
+		}
+		// body
+		if (blockRequest.RequestedData&2)>>1 == 1 {
+			retData, err := s.blockState.GetBlockBody(hash)
+			if err == nil && retData != nil {
+				blockData.Body = retData.AsOptional()
+			}
+		}
+		// receipt
+		if (blockRequest.RequestedData&4)>>2 == 1 {
+			retData, err := s.blockState.GetReceipt(hash)
+			if err == nil && retData != nil {
+				blockData.Receipt = optional.NewBytes(true, retData)
+			}
+		}
+		// message queue
+		if (blockRequest.RequestedData&8)>>3 == 1 {
+			retData, err := s.blockState.GetMessageQueue(hash)
+			if err == nil && retData != nil {
+				blockData.MessageQueue = optional.NewBytes(true, retData)
+			}
+		}
+		// justification
+		if (blockRequest.RequestedData&16)>>4 == 1 {
+			retData, err := s.blockState.GetJustification(hash)
+			if err == nil && retData != nil {
+				blockData.Justification = optional.NewBytes(true, retData)
+			}
+		}
+
+		responseData = append(responseData, blockData)
+	}
+
+	return &network.BlockResponseMessage{
+		ID:        blockRequest.ID,
+		BlockData: responseData,
+	}, nil
+}
+
+// BlockResponseMessage 2
+
+// ProcessBlockResponseMessage attempts to validate and add the block to the
+// chain by calling `core_execute_block`. Valid blocks are stored in the block
+// database to become part of the canonical chain.
+func (s *Service) ProcessBlockResponseMessage(msg *network.BlockResponseMessage) error {
+	log.Debug("[core] received BlockResponseMessage")
+
+	// loop through block data in block response message
+	for _, bd := range msg.BlockData {
+		if bd.Header.Exists() {
+
+			header, err := types.NewHeader(
+				bd.Header.Value().ParentHash,
+				bd.Header.Value().Number,
+				bd.Header.Value().StateRoot,
+				bd.Header.Value().ExtrinsicsRoot,
+				bd.Header.Value().Digest,
+			)
+			if err != nil {
+				return err
+			}
+
+			// if authority, check if first block and set epoch data for next epoch
+			if s.isAuthority {
+				err = s.handleConsensusDigest(header)
```

Comment:
what about the case when firstBlock is already set?

Reply:
I think you should change this statement to only execute is firstBlock for the epoch is not already set, *or* if a block is received with a lower number that the current firstBlock but is still in the epoch

Reply:
it is inside `handleConsensusDigest` because you request that this be consolidated